### PR TITLE
External Authentication

### PR DESCRIPTION
**Pull request by wemrysi**
Adds support for bearer token authentication using the quasar `ExternalCredentials` provider.

[ch11533]

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -29,7 +29,6 @@ lazy val root = project
 
 lazy val core = project
   .in(file("core"))
-  .settings(addCompilerPlugin("com.olegpy" %% "better-monadic-for" % "0.3.1"))
```
Comment (alissapajer) [inline build.sbt:32]: 👍

**Comment by wemrysi** [comment]
`breaking` as we're enforcing mutual exclusivity of `reqTokAuth`, `oauth2` and `externalAuth` configuration.

File: build.sbt [hunk 2]
```diff
@@ -52,15 +51,18 @@ lazy val core = project
...
       "com.precog" %% "http4s-argonaut" % http4sVersion,
       "org.http4s" %% "jawn-fs2" % "1.0.0-RC2",
-      "com.precog" %% "http4s-dsl" % http4sVersion % Test,
       "com.codecommit" %% "shims" % "2.0.0",
       "com.typesafe.akka" %% "akka-http" % "10.1.9",
       "com.typesafe.akka" %% "akka-stream" % "2.5.25",
       "org.scodec" %% "scodec-core" % "1.10.3",
+      "com.precog" %% "quasar-impl" % quasarVersion.value),
+
+    libraryDependencies ++= Seq(
```
Comment (alissapajer) [inline build.sbt:60]: 👍 thanks!

File: core/src/test/scala/quasar/datasource/url/ReconfigurationSpec.scala [hunk 3]
```diff
@@ -175,6 +175,43 @@ object ReconfigurationSpec extends Specification {
...
+      val patchSanitized = Json(
+        "maxRedirects" := 5,
+        "requests" := List[Json](),
+        "format" := ldjsonJ,
+        "headers" := Json(
+          "Accept" := "*/*",
+          "connection" := "Upgrade"),
+        "externalAuth" := Json(
+          "type" := "bearer-token",
+          "authId" := "00000000-0000-0000-0000-000000000000",
```
Comment (alissapajer) [inline core/src/test/scala/quasar/datasource/url/ReconfigurationSpec.scala:204]: 👍

File: core/src/test/scala/quasar/datasource/url/ExternalCredentialsMiddlewareSpec.scala [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  val scheme = CIS("Token")
+  val endpoint = Uri.uri("http://example.com/data")
+  val config = ExternalAuthConfig.BearerTokenConfig(UUID.randomUUID, scheme)
+
+  def unauthorized(msg: String): IO[Response[IO]] =
+    Unauthorized(`WWW-Authenticate`(Challenge(scheme.value, "example.com")))
+      .map(_.withEntity(msg))
+
+  def client(creds: ExternalCredentials[IO], routes: HttpRoutes[IO]): Client[IO] =
+    ExternalCredentialsMiddleware(creds, config)(Client.fromHttpApp(routes.orNotFound))
```
Comment (alissapajer) [inline core/src/test/scala/quasar/datasource/url/ExternalCredentialsMiddlewareSpec.scala:71]: I hadn't fully appreciated how nice it is to be able to test a middleware component as a `Client`!

**Comment by cryogenian** [comment]
Wait

**Comment by cryogenian** [comment]
Only parts of OAuth configs and ReqToken configs are used as actual source of authorization. Some of them are needed at frontend. Suppose, I set scopes to be `foo, bar, baz` and grabbed auth bearer, w/o preserving scopes at web-ds config I wouldn't be able  to determine what scopes were set during editing. 

The same about request token authorization. Having `path` in config could make ds modification much much simpler

**Pull request merged**